### PR TITLE
[Documentation] Developers guide for TSDB migration

### PR DESCRIPTION
**Pull request by agithomas**
Type of Change
- Enhancement

## What does this PR do?

Adding documentation of : A Developers guide for TSDB migration

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

### CODE REVIEW

File: docs/developer_tsdb_migration_guidelines.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+# <a id="background"></a> Background
+
+A time series is a sequence of observations for a specific entity. TSDB enables the column-oriented functionality in elasticsearch by co-locating the data and optimizing the storage and aggregations to take advantage of such co-allocation.
+
+Integration is one of the biggest sources of input data to elasticsearch. Enabling TSDB on integration packages can be achieved by minimal changes made in `fields.yml` and `manifest.yml` files of a package.
+
+# <a id="existing-migrated-packages"></a> Existing Packages Already Migrated
+
+Oracle Package (Meta Issue) : [PR Link](https://github.com/elastic/observability-dev/issues/2541)
```
Comment (lalit-satapathy) [inline docs/developer_tsdb_migration_guidelines.md:21]: We can link to the public Oracle PR (latest one) not the private link

File: docs/developer_tsdb_migration_guidelines.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    A field having type [flattened](https://www.elastic.co/guide/en/elasticsearch/reference/current/flattened.html) cannot be selected as a dimension field. If the field that you are choosing as a dimension is too long or is of type flattened , consider the option of hashing the value of this field, creating a new dimension field to hold this value  . [Fingerprint processor](https://www.elastic.co/guide/en/elasticsearch/reference/current/fingerprint-processor.html) can be used for this purpose.  
+    
+    Reference :  [Oracle Integration TSDB Enablement Example](https://github.com/elastic/integrations/blob/8a57d6ba96d391afc33da20c80ec51280d22f009/packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml#LL127C4-L131C29)  
+
+    From the context of packages having ownership of the service integration team, there exist certain fields that are part of every package and they are potential candidates of becoming dimension fields
```
Comment (lalit-satapathy) [inline docs/developer_tsdb_migration_guidelines.md:49]: Let's remove reference to "service integration team"

File: docs/developer_tsdb_migration_guidelines.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    When metrics are collected from a resource running in the cloud or in a container, certain fields are potential candidates of becoming dimension fields  
+
+    * host.ip
+    * service.address
+    * agent.id
+    * cloud.project.id
+    * cloud.instance.id
+    * cloud.provider
+    * container.id  
+
```
Comment (lalit-satapathy) [inline docs/developer_tsdb_migration_guidelines.md:64]: Need a clarification that these are ECS common fields.

File: docs/developer_tsdb_migration_guidelines.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+Verify that all the dashboards related to the datastream are showing the data.  
+
+- If a non-Lens visualization is used, verify if the dashboard is rendering the data properly. Certain aggregation functions are not supported when a field is having a metric_type ‘counter’. Example avg(). Replace such aggregation functions with a supported aggregation type such as max()  
+
+- It is recommended to compare the number of documents within a certain time frame before enabling the TSDB and after enabling TSDB index mode. If the count differs, please check if there exists a field that is not annotated as dimension field.  
+
+
+# <a id="best-practices"></a> Best Practices
+
+- [Lens](https://www.elastic.co/guide/en/kibana/current/lens.html) must be the default visualization type that must be used. If the visualization type is not Lens, migrate to Lens visualization type first.
```
Comment (lalit-satapathy) [inline docs/developer_tsdb_migration_guidelines.md:118]: Lens migration should not be mandatory

**Review by lalit-satapathy — COMMENTED** [review]
Added comments

File: docs/developer_tsdb_migration_guidelines.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+Oracle Package (Meta Issue) : [PR Link](https://github.com/elastic/observability-dev/issues/2541)  
+
+# <a id="migration-steps"></a> Steps for Migrating an existing Package
+
+
+1. **Datastream having type`logs`can be excluded from TSDB migration.**
+2. **Add the changes to the manifest.yml file of the datastream as below to enable the timeseries index mode**
+    ```
+    elasticsearch:
+    index_mode: "time_series"
```
Comment (jsoriano) [inline docs/developer_tsdb_migration_guidelines.md:30]: Indentation here.
```suggestion
      index_mode: "time_series"
```

File: docs/developer_tsdb_migration_guidelines.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    ```
+    If your datastream has more number of dimension fields, you can modify this limit by modifying index.mapping.dimension_fields.limit value as below
+    ```
+    elasticsearch:
+    index_mode: "time_series"
+    index_template:
+    settings:
+      # Defaults to 16
+      index.mapping.dimension_fields.limit: 32
```
Comment (jsoriano) [inline docs/developer_tsdb_migration_guidelines.md:39]: Indentation:
```suggestion
      index_mode: "time_series"
      index_template:
        settings:
          # Defaults to 16
          index.mapping.dimension_fields.limit: 32
```

File: docs/developer_tsdb_migration_guidelines.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    ```
+3. **Identifying the dimensions in the datastream.** 
+
+    Read about dimension fields [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/tsds.html#time-series-dimension). It is important that dimensions or a set of dimensions that are part of a datastream uniquely identify a timeseries. Dimensions are used to form _tsid which then is used for routing and index sorting.
```
Comment (jsoriano) [inline docs/developer_tsdb_migration_guidelines.md:43]: Something like this should be included in the general guidelines too, not only in this migration guide. This can be done in a separate PR like https://github.com/elastic/integrations/pull/5270, but this content could be linked here instead of duplicating it.

File: docs/developer_tsdb_migration_guidelines.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+    A field having type [flattened](https://www.elastic.co/guide/en/elasticsearch/reference/current/flattened.html) cannot be selected as a dimension field. If the field that you are choosing as a dimension is too long or is of type flattened , consider the option of hashing the value of this field, creating a new dimension field to hold this value  . [Fingerprint processor](https://www.elastic.co/guide/en/elasticsearch/reference/current/fingerprint-processor.html) can be used for this purpose.
```
Comment (jsoriano) [inline docs/developer_tsdb_migration_guidelines.md:45]: There are many only some types that can be selected as dimensions. This should be also part of the general guidelines. There is actually a mention to this already there https://github.com/elastic/integrations/blob/main/docs/generic_guidelines.md#specify-dimensions.

It is great though to provide instructions about what to do with certain types, as is being done here for the `flattened` type.

File: docs/developer_tsdb_migration_guidelines.md [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+    * host.ip
+    * service.address
+    * agent.id
+    * cloud.project.id
+    * cloud.instance.id
+    * cloud.provider
+    * container.id  
+
+    *Warning: Choosing an insufficient number of dimension fields may lead to data loss*
```
Comment (jsoriano) [inline docs/developer_tsdb_migration_guidelines.md:65]: This should be mentioned also in the general guidelines, it could include a link there https://github.com/elastic/integrations/pull/5270/files#diff-04ec6b271964b116e3aa8a0f59b1f3325ae14bbe8309fac40887b0ef9c782cabR62

File: docs/developer_tsdb_migration_guidelines.md (hunk 4 above)
Comment (jsoriano) [reply docs/developer_tsdb_migration_guidelines.md:118]: Agree, I would rephrase to mean something like: "Test the dashboards after migrating to TSDB, if some visualization doesn't work, consider migrating it to Lens".

**Review by jsoriano — COMMENTED** [review]
A lot of useful information here, thanks!

It focus on the migration, but most of the information here would be useful also when creating new integrations, or just to know more about the use of TSDB.

I would propose to:
* Add something to the general guidelines, https://github.com/elastic/integrations/pull/5270 or similar.
* Add a `tsdb_guidelines.md` document with more details, including most of the information here, but targeting also developers of new packages. For example this document could contain information about what to do with certain unsupported types, as the workaround mentioned here for the `flattened` type. Or the specific code snippets.
* Add a migration guideline as an specific section inside this `tsdb_guidelines.md` doc, it could reference other parts of the document.

File: docs/developer_tsdb_migration_guidelines.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+An enhancement [request](https://github.com/elastic/kibana/issues/150549) for Kibana is created to indicate the write index. Until then, refer to the index.time_series.start_time of indices and compare with the current time to identify the write index. 
+
+*Hint: In the Index Management UI, against a specific index, if the  docs count column values regularly increase for an Index, it can be considered as the write index*
+
+**Conflicting Field Type** : Fields having conflicting field type will not be considered as dimension. Resolve the field type ambiguity before defining a field as dimension field.
+
+# <a id="known-issues"></a> Known Issues
+
+- Lens visualization fails if a field is having a metric_type value as ‘counter’ and certain aggregation functions are applied over it.
```
Comment (ishleenk17) [inline docs/developer_tsdb_migration_guidelines.md:145]: Whys should a lens related issues be reported in TSDB documentation ?

File: docs/developer_tsdb_migration_guidelines.md [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    * *There exists a limit on how many dimension fields can have. By default this value is 16. Out of this, 8 are reserved for ecs fields.*
+
+    * *Dimension keys have a hard limit of 512b. Documents are rejected if this limit is reached.*
+
+    * *Dimension values have a hard limit of 1024b. Documents are rejected if this limit is reached*
+
+
+5. **Annotating Metric Types values for all applicable fields** 
+
+    Metrics are fields that contain numeric measurements, as well as aggregations and/or downsampling values based off of those measurements.
```
Comment (ishleenk17) [inline docs/developer_tsdb_migration_guidelines.md:88]: Is having a metric_type relevant to dimensions, tsdb ?

Comment (jsoriano) [reply docs/developer_tsdb_migration_guidelines.md:88]: Not really needed for TSDB, but in the TSDB use case helps to give more information to Kibana about how to represent and query the data.

**Comment by lalit-satapathy** [comment]
> A lot of useful information here, thanks!
>It focus on the migration, but most of the information here would be useful also when creating new integrations, or just to know more about the use of TSDB.

Agree the [generic guideline](https://github.com/elastic/integrations/pull/5270)  has some very useful dimension information, needed for any package developers.  @jsoriano, let's commit that PR soon.

Where this guideline [document](https://github.com/elastic/integrations/blob/1b048f8e3849e5eabcb07419d3df54d60a7fa264/docs/developer_tsdb_migration_guidelines.md), is needed is particularly for migrating multiple existing packages in short term. @agithomas let's close this soon.

File: docs/developer_tsdb_migration_guidelines.md (hunk 10 above)
Comment (agithomas) [reply docs/developer_tsdb_migration_guidelines.md:149]: It is not a Lens issue.

File: docs/developer_tsdb_migration_guidelines.md (hunk 8 above)
Comment (agithomas) [reply docs/developer_tsdb_migration_guidelines.md:45]: Linked to the appropriate section in the generic_guidelines.md document

**Comment by agithomas** [comment]
@jsoriano , can you please do a final review of the PR, please?

**Review by lalit-satapathy — APPROVED** [review]
@jsoriano, Can you merge this document? 

We need to continue updating this doc regularly and in quick time, as more feedbacks come during the migration process.

**Review by jsoriano — APPROVED** [review]
